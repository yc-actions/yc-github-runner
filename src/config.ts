import * as core from '@actions/core'
import * as github from '@actions/github'
import { parseMemory } from './memory'
import moment from 'moment'

export interface ResourcesSpec {
    memory: number
    cores: number
    coreFraction: number
}

export interface ActionConfig {
    imageId: string
    mode: string
    githubToken: string
    runnerHomeDir: string
    label: string
    subnetId: string
    publicIp: boolean
    serviceAccountId: string
    diskType: string
    diskSize: number
    folderId: string
    zoneId: string
    platformId: string
    resourcesSpec: ResourcesSpec

    secondDiskImageId: string
    secondDiskType: string
    secondDiskSize: number

    user: string
    sshPublicKey: string

    instanceId?: string

    runnerVersion: string
    ttl?: moment.Duration
    disableUpdate: boolean
}

export interface GithubRepo {
    owner: string
    repo: string
}

export class Config {
    input: ActionConfig
    githubContext: GithubRepo

    constructor(input?: ActionConfig) {
        this.input = input ?? parseVmInputs()

        // the values of github.context.repo.owner and github.context.repo.repo are taken from
        // the environment variable GITHUB_REPOSITORY specified in "owner/repo" format and
        // provided by the GitHub Action on the runtime
        this.githubContext = {
            owner: github.context.repo.owner,
            repo: github.context.repo.repo
        }

        //
        // validate input
        //

        if (!this.input.mode) {
            throw new Error(`The 'mode' input is not specified`)
        }

        if (!this.input.githubToken) {
            throw new Error(`The 'github-token' input is not specified`)
        }

        if (this.input.mode === 'start') {
            if (!this.input.imageId || !this.input.subnetId || !this.input.folderId) {
                throw new Error(`Not all the required inputs are provided for the 'start' mode`)
            }

            if (this.input.secondDiskSize > 0 && !this.input.secondDiskImageId) {
                throw new Error(`Secondary disk image id is missing`)
            }
        } else if (this.input.mode === 'stop') {
            if (!this.input.label || !this.input.instanceId) {
                throw new Error(`Not all the required inputs are provided for the 'stop' mode`)
            }
        } else {
            throw new Error('Wrong mode. Allowed values: start, stop.')
        }
    }

    generateUniqueLabel(): string {
        return Math.random().toString(36).slice(2, 7)
    }
}

function parseVmInputs(): ActionConfig {
    core.startGroup('Parsing Action Inputs')

    const folderId: string = core.getInput('folder-id')

    const mode = core.getInput('mode')
    const githubToken = core.getInput('github-token')
    const runnerHomeDir = core.getInput('runner-home-dir')
    const label = core.getInput('label')

    const serviceAccountId: string = core.getInput('service-account-id')

    const imageId: string = core.getInput('image-id')
    const zoneId: string = core.getInput('zone-id') || 'ru-central1-a'
    const subnetId: string = core.getInput('subnet-id')
    const publicIp: boolean = core.getBooleanInput('public-ip', { required: false })
    const platformId: string = core.getInput('platform-id') || 'standard-v3'
    const cores: number = parseInt(core.getInput('cores') || '2', 10)
    const memory: number = parseMemory(core.getInput('memory') || '1Gb')
    const diskType: string = core.getInput('disk-type') || 'network-ssd'
    const diskSize: number = parseMemory(core.getInput('disk-size') || '30Gb')
    const coreFraction: number = parseInt(core.getInput('core-fraction') || '100', 10)

    const secondDiskImageId: string = core.getInput('image2-id')
    const secondDiskType: string = core.getInput('disk2-type') || 'network-ssd'
    const secondDiskSize: number = parseMemory(core.getInput('disk2-size') || '0Gb')

    const user: string = core.getInput('user')
    const sshPublicKey: string = core.getInput('ssh-public-key')

    const instanceId: string = core.getInput('instance-id', { required: false })

    const runnerVersion: string = core.getInput('runner-version', { required: false })
    const disableUpdate: boolean = core.getBooleanInput('disable-update', { required: false })

    let ttl: moment.Duration | undefined = undefined
    const ttlInput = core.getInput('ttl', { required: false })
    if (ttlInput) {
        ttl = moment.duration(ttlInput)
    }

    core.endGroup()
    return {
        instanceId,
        imageId,
        diskType,
        diskSize,
        subnetId,
        publicIp,
        zoneId,
        platformId,
        folderId,
        mode,
        githubToken,
        runnerHomeDir,
        label,
        serviceAccountId,
        secondDiskImageId,
        secondDiskType,
        secondDiskSize,
        user,
        sshPublicKey,
        runnerVersion,
        ttl,
        disableUpdate,
        resourcesSpec: {
            cores,
            memory,
            coreFraction
        }
    }
}
