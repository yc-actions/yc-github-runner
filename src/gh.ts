import { info, error as core_error } from '@actions/core'
import { getOctokit } from '@actions/github'
import { Config } from './config'

interface Runner {
    /** The id of the runner. */
    id: number
    /** The name of the runner. */
    name: string
    /** The Operating System of the runner. */
    os: string
    /** The status of the runner. */
    status: string
    busy: boolean
    labels: Array<{
        /** Unique identifier of the label. */
        id?: number
        /** Name of the label. */
        name?: string
        /** The type of label. Read-only labels are applied automatically when the runner is configured. */
        type?: 'read-only' | 'custom'
    }>
}

// use the unique label to find the runner
// as we don't have the runner's id, it's not possible to get it in any other way
export async function getRunner(config: Config, label: string): Promise<Runner | null> {
    const octokit = getOctokit(config.input.githubToken)

    try {
        const runners = (await octokit.paginate(
            'GET /repos/{owner}/{repo}/actions/runners',
            config.githubContext
        )) as Runner[]
        const foundRunners = runners.filter(x => x.labels.some(l => l.name === label))
        return foundRunners.length > 0 ? foundRunners[0] : null
    } catch (error) {
        info(`GitHub self-hosted runner receiving error: ${error}`)
        return null
    }
}

// get GitHub Registration Token for registering a self-hosted runner
export async function getRegistrationToken(config: Config): Promise<string> {
    const octokit = getOctokit(config.input.githubToken)

    try {
        const req = {
            ...config.githubContext
        }
        const response = await octokit.request('POST /repos/{owner}/{repo}/actions/runners/registration-token', req)
        info('GitHub Registration Token is received')
        return response.data.token
    } catch (error) {
        core_error('GitHub Registration Token receiving error')
        throw error
    }
}

export async function removeRunner(config: Config): Promise<void> {
    const runner = await getRunner(config, config.input.label)
    const octokit = getOctokit(config.input.githubToken)

    // skip the runner removal process if the runner is not found
    if (!runner) {
        info(`GitHub self-hosted runner with label ${config.input.label} is not found, so the removal is skipped`)
        return
    }

    try {
        const req = {
            ...config.githubContext,
            runner_id: runner.id
        }
        await octokit.request('DELETE /repos/{owner}/{repo}/actions/runners/{runner_id}', req)
        info(`GitHub self-hosted runner ${runner.name} is removed`)
        return
    } catch (error) {
        core_error('GitHub self-hosted runner removal error')
        throw error
    }
}

export async function waitForRunnerRegistered(config: Config, label: string): Promise<void> {
    const timeoutMinutes = 5
    const retryIntervalSeconds = 10
    const quietPeriodSeconds = 30
    let waitSeconds = 0

    info(`Waiting ${quietPeriodSeconds}s for the instance to be registered in GitHub as a new self-hosted runner`)
    await new Promise(r => setTimeout(r, quietPeriodSeconds * 1000))
    info(`Checking every ${retryIntervalSeconds}s if the GitHub self-hosted runner is registered`)

    return new Promise((resolve, reject) => {
        const interval = setInterval(async () => {
            const runner = await getRunner(config, label)

            if (waitSeconds > timeoutMinutes * 60) {
                core_error('GitHub self-hosted runner registration error')
                clearInterval(interval)

                reject(
                    new Error(
                        `A timeout of ${timeoutMinutes} minutes is exceeded. Your YC instance was not able to register itself in GitHub as a new self-hosted runner.`
                    )
                )
            }

            if (runner && runner.status === 'online') {
                info(`GitHub self-hosted runner ${runner.name} is registered and ready to use`)
                clearInterval(interval)
                resolve()
            } else {
                waitSeconds += retryIntervalSeconds
                info('Checking...')
            }
        }, retryIntervalSeconds * 1000)
    })
}
