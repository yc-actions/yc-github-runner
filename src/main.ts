import * as core from '@actions/core'
import * as github from '@actions/github'
import {
    decodeMessage,
    errors,
    serviceClients,
    Session,
    waitForOperation,
    WrappedServiceClientType
} from '@yandex-cloud/nodejs-sdk'
import { Instance, IpVersion } from '@yandex-cloud/nodejs-sdk/dist/generated/yandex/cloud/compute/v1/instance'
import {
    AttachedDiskSpec,
    AttachedDiskSpec_Mode,
    CreateInstanceMetadata,
    CreateInstanceRequest,
    DeleteInstanceRequest,
    InstanceServiceService,
    NetworkInterfaceSpec
} from '@yandex-cloud/nodejs-sdk/dist/generated/yandex/cloud/compute/v1/instance_service'
import * as yaml from 'js-yaml'
import { Config, GithubRepo } from './config'
import { getRegistrationToken, removeRunner, waitForRunnerRegistered } from './gh'
import { fromServiceAccountJsonFile } from './service-account-json'
import moment from 'moment'

let config: Config

try {
    config = new Config()
} catch (error) {
    const err = error as Error
    core.error(err)
    core.setFailed(err.message)
}

interface BuildUserDataScriptParams {
    githubRegistrationToken: string
    label: string
    runnerHomeDir: string
    owner: string
    repo: string
    user: string
    sshPublicKey: string
    runnerVersion: string
}

// User data scripts are run as the root user
export function buildUserDataScript(params: BuildUserDataScriptParams): string[] {
    const { githubRegistrationToken, label, runnerHomeDir, repo, owner, user, sshPublicKey } = params
    let script: string[]
    /*eslint-disable max-len*/
    if (runnerHomeDir) {
        // If runner home directory is specified, we expect the actions-runner software (and dependencies)
        // to be pre-installed in the image, so we simply cd into that directory and then start the runner
        script = [
            '#!/bin/bash',
            `cd "${runnerHomeDir}"`,
            'export RUNNER_ALLOW_RUNASROOT=1',
            `./config.sh --url https://github.com/${owner}/${repo} --token ${githubRegistrationToken} --labels ${label}`,
            './run.sh'
        ]
    } else {
        const version = params.runnerVersion
        script = [
            '#!/bin/bash',
            'mkdir actions-runner && cd actions-runner',
            'case $(uname -m) in aarch64) ARCH="arm64" ;; amd64|x86_64) ARCH="x64" ;; esac && export RUNNER_ARCH=${ARCH}',
            `curl -O -L https://github.com/actions/runner/releases/download/v${version}/actions-runner-linux-\${RUNNER_ARCH}-${version}.tar.gz`,
            `tar xzf ./actions-runner-linux-\${RUNNER_ARCH}-${version}.tar.gz`,
            'export RUNNER_ALLOW_RUNASROOT=1',
            `./config.sh --url https://github.com/${owner}/${repo} --token ${githubRegistrationToken} --labels ${label}`,
            './run.sh'
        ]
    }
    if (user !== '' && sshPublicKey !== '') {
        const cloudInit = yaml.load(`ssh_pwauth: no
users:
  - name: ${user}
    sudo: ALL=(ALL) NOPASSWD:ALL
    shell: /bin/bash
    ssh_authorized_keys:
      - "${sshPublicKey}"`) as Record<string, unknown>
        cloudInit['runcmd'] = script.slice(1)
        return ['#cloud-config', ...yaml.dump(cloudInit).split('\n')]
    } else {
        return script
    }
}

async function createVm(
    session: Session,
    instanceService: WrappedServiceClientType<typeof InstanceServiceService>,
    repo: GithubRepo,
    githubRegistrationToken: string,
    label: string
): Promise<string> {
    core.startGroup('Create VM')

    const secondaryDiskSpecs: AttachedDiskSpec[] = []

    if (config.input.secondDiskSize > 0) {
        secondaryDiskSpecs.push(
            AttachedDiskSpec.fromPartial({
                autoDelete: true,
                mode: AttachedDiskSpec_Mode.READ_WRITE,
                diskSpec: {
                    imageId: config.input.secondDiskImageId,
                    size: config.input.secondDiskSize,
                    typeId: config.input.secondDiskType
                }
            })
        )
    }
    let primaryV4AddressSpec = {}

    if (config.input.publicIp) {
        primaryV4AddressSpec = {
            oneToOneNatSpec: {
                ipVersion: IpVersion.IPV4
            }
        }
    }

    const networkInterfaceSpec = NetworkInterfaceSpec.fromPartial({
        subnetId: config.input.subnetId,
        primaryV4AddressSpec
    })

    const labels: Record<string, string> = {}

    if (config.input.ttl) {
        // Set `expires` label to the current time + TTL Duration
        // Instance won't automatically be destroyed by Yandex.Cloud, you should handle it yourself
        // For example, by using Cron trigger that will call Cloud Function to destroy the instance.
        labels['expires'] = moment.utc().add(config.input.ttl).unix().toString()
    }

    const op = await instanceService.create(
        CreateInstanceRequest.fromPartial({
            folderId: config.input.folderId,
            description: `Runner for: ${repo.owner}/${repo.repo}`,
            zoneId: config.input.zoneId,
            platformId: config.input.platformId,
            resourcesSpec: config.input.resourcesSpec,
            metadata: {
                'user-data': buildUserDataScript({
                    githubRegistrationToken,
                    label,
                    runnerHomeDir: config.input.runnerHomeDir,
                    user: config.input.user,
                    sshPublicKey: config.input.sshPublicKey,
                    repo: config.githubContext.repo,
                    owner: config.githubContext.owner,
                    runnerVersion: config.input.runnerVersion
                }).join('\n')
            },
            labels,

            bootDiskSpec: {
                mode: AttachedDiskSpec_Mode.READ_WRITE,
                autoDelete: true,
                diskSpec: {
                    typeId: config.input.diskType,
                    size: config.input.diskSize,
                    imageId: config.input.imageId
                }
            },
            secondaryDiskSpecs,
            networkInterfaceSpecs: [networkInterfaceSpec],
            serviceAccountId: config.input.serviceAccountId
        })
    )
    const finishedOp = await waitForOperation(op, session)
    if (finishedOp.response) {
        const instanceId = decodeMessage<Instance>(finishedOp.response).id
        core.info(`Created instance with id '${instanceId}'`)
        core.endGroup()
        return instanceId
    } else {
        core.error(`Failed to create instance'`)
        core.endGroup()
        throw new Error('Failed to create instance')
    }
}

async function destroyVm(
    session: Session,
    instanceService: WrappedServiceClientType<typeof InstanceServiceService>
): Promise<void> {
    core.startGroup('Create VM')

    const op = await instanceService.delete(
        DeleteInstanceRequest.fromPartial({
            instanceId: config.input.instanceId
        })
    )
    const finishedOp = await waitForOperation(op, session)
    if (finishedOp.metadata) {
        const instanceId = decodeMessage<CreateInstanceMetadata>(finishedOp.metadata).instanceId
        core.info(`Destroyed instance with id '${instanceId}'`)
    } else {
        core.error(`Failed to create instance'`)
        throw new Error('Failed to create instance')
    }
    core.endGroup()
}

async function start(
    session: Session,
    instanceService: WrappedServiceClientType<typeof InstanceServiceService>
): Promise<void> {
    const label = config.generateUniqueLabel()
    const githubRegistrationToken = await getRegistrationToken(config)
    const instanceId = await createVm(session, instanceService, github.context.repo, githubRegistrationToken, label)
    core.setOutput('label', label)
    core.setOutput('instance-id', instanceId)
    await waitForRunnerRegistered(config, label)
}

async function stop(
    session: Session,
    instanceService: WrappedServiceClientType<typeof InstanceServiceService>
): Promise<void> {
    await destroyVm(session, instanceService)
    await removeRunner(config)
}

async function run(): Promise<void> {
    core.setCommandEcho(true)
    try {
        core.info(`start`)
        const ycSaJsonCredentials = core.getInput('yc-sa-json-credentials', {
            required: true
        })

        core.info(`Folder ID: ${config.input.folderId}`)

        const serviceAccountJson = fromServiceAccountJsonFile(JSON.parse(ycSaJsonCredentials))
        core.info('Parsed Service account JSON')

        const session = new Session({ serviceAccountJson })
        const instanceService = session.client(serviceClients.InstanceServiceClient)

        switch (config.input.mode) {
            case 'start': {
                await start(session, instanceService)
                break
            }
            case 'stop': {
                await stop(session, instanceService)
                break
            }
            default:
                // noinspection ExceptionCaughtLocallyJS
                throw new Error(`Unknown mode ${config.input.mode}`)
        }
    } catch (error) {
        if (error instanceof errors.ApiError) {
            core.error(`${error.message}\nx-request-id: ${error.requestId}\nx-server-trace-id: ${error.serverTraceId}`)
        }
        core.setFailed(error as Error)
    }
}

run()
