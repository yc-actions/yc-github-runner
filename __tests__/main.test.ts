import { describe, expect, test } from '@jest/globals'
import * as cp from 'child_process'
import * as path from 'path'
import * as process from 'process'
import { buildUserDataScript } from '../src/main'

// This test will run only in fully configured env and creates real VM
// in the Yandex Cloud, so it will be disabled in CI/CD. You can enable it to test locally.
test.skip('runs', () => {
    process.env['GITHUB_WORKSPACE'] = ''

    const np = process.execPath
    const ip = path.join(__dirname, '..', 'lib', 'main.js')
    const options: cp.ExecFileSyncOptions = {
        env: process.env
    }
    let res
    try {
        res = cp.execFileSync(np, [ip], options)
    } catch (e) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        console.log((e as any).stdout.toString())
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        console.log((e as any).stderr.toString())
    }
    console.log(res?.toString())
})

describe('cloud-init', () => {
    test('default', () => {
        const actual = buildUserDataScript({
            githubRegistrationToken: 'githubRegistrationToken',
            label: 'label',
            owner: 'owner',
            repo: 'repo',
            runnerHomeDir: '',
            user: '',
            sshPublicKey: '',
            runnerVersion: '2.299.1',
            disableUpdate: false
        })
        expect(actual.length).toBe(8)
        expect(actual).toMatchSnapshot()
        expect(actual[0]).toBe('#!/bin/bash')
    })

    test('with home dir', () => {
        const actual = buildUserDataScript({
            githubRegistrationToken: 'githubRegistrationToken',
            label: 'label',
            owner: 'owner',
            repo: 'repo',
            runnerHomeDir: 'foo',
            user: '',
            sshPublicKey: '',
            runnerVersion: '2.299.1',
            disableUpdate: false
        })
        expect(actual.length).toBe(5)
        expect(actual).toMatchSnapshot()
        expect(actual[0]).toBe('#!/bin/bash')
    })

    test('with user and ssh key', () => {
        const actual = buildUserDataScript({
            githubRegistrationToken: 'githubRegistrationToken',
            label: 'label',
            owner: 'owner',
            repo: 'repo',
            runnerHomeDir: '',
            user: 'user',
            sshPublicKey: 'key',
            runnerVersion: '2.299.1',
            disableUpdate: false
        })
        expect(actual.length).toBe(23)
        expect(actual).toMatchSnapshot()
        expect(actual[0]).toBe('#cloud-config')
    })

    test('with home dir and user and ssh key', () => {
        const actual = buildUserDataScript({
            githubRegistrationToken: 'githubRegistrationToken',
            label: 'label',
            owner: 'owner',
            repo: 'repo',
            runnerHomeDir: 'foo',
            user: 'user',
            sshPublicKey: 'key',
            runnerVersion: '2.299.1',
            disableUpdate: false
        })
        expect(actual.length).toBe(16)
        expect(actual).toMatchSnapshot()
        expect(actual[0]).toBe('#cloud-config')
    })

    test('with disable update', () => {
        const actual = buildUserDataScript({
            githubRegistrationToken: 'githubRegistrationToken',
            label: 'label',
            owner: 'owner',
            repo: 'repo',
            runnerHomeDir: '',
            user: '',
            sshPublicKey: '',
            runnerVersion: '2.299.1',
            disableUpdate: true
        })
        expect(actual.length).toBe(8)
        expect(actual).toMatchSnapshot()
        expect(actual[6]).toContain('--disableupdate')
    })

    test('without disable update', () => {
        const actual = buildUserDataScript({
            githubRegistrationToken: 'githubRegistrationToken',
            label: 'label',
            owner: 'owner',
            repo: 'repo',
            runnerHomeDir: '',
            user: '',
            sshPublicKey: '',
            runnerVersion: '2.299.1',
            disableUpdate: false
        })
        expect(actual.length).toBe(8)
        expect(actual).toMatchSnapshot()
        expect(actual[6]).not.toContain('--disableupdate')
    })
})
