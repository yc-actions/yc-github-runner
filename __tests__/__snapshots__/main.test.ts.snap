// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cloud-init default 1`] = `
[
  "#!/bin/bash",
  "mkdir actions-runner && cd actions-runner",
  "case $(uname -m) in aarch64) ARCH="arm64" ;; amd64|x86_64) ARCH="x64" ;; esac && export RUNNER_ARCH=\${ARCH}",
  "curl -O -L https://github.com/actions/runner/releases/download/v2.299.1/actions-runner-linux-\${RUNNER_ARCH}-2.299.1.tar.gz",
  "tar xzf ./actions-runner-linux-\${RUNNER_ARCH}-2.299.1.tar.gz",
  "export RUNNER_ALLOW_RUNASROOT=1",
  "./config.sh --url https://github.com/owner/repo --token githubRegistrationToken --labels label",
  "./run.sh",
]
`;

exports[`cloud-init with disable update 1`] = `
[
  "#!/bin/bash",
  "mkdir actions-runner && cd actions-runner",
  "case $(uname -m) in aarch64) ARCH="arm64" ;; amd64|x86_64) ARCH="x64" ;; esac && export RUNNER_ARCH=\${ARCH}",
  "curl -O -L https://github.com/actions/runner/releases/download/v2.299.1/actions-runner-linux-\${RUNNER_ARCH}-2.299.1.tar.gz",
  "tar xzf ./actions-runner-linux-\${RUNNER_ARCH}-2.299.1.tar.gz",
  "export RUNNER_ALLOW_RUNASROOT=1",
  "./config.sh --url https://github.com/owner/repo --token githubRegistrationToken --labels label --disableupdate",
  "./run.sh",
]
`;

exports[`cloud-init with home dir 1`] = `
[
  "#!/bin/bash",
  "cd "foo"",
  "export RUNNER_ALLOW_RUNASROOT=1",
  "./config.sh --url https://github.com/owner/repo --token githubRegistrationToken --labels label",
  "./run.sh",
]
`;

exports[`cloud-init with home dir and user and ssh key 1`] = `
[
  "#cloud-config",
  "ssh_pwauth: 'no'",
  "users:",
  "  - name: user",
  "    sudo: ALL=(ALL) NOPASSWD:ALL",
  "    shell: /bin/bash",
  "    ssh_authorized_keys:",
  "      - key",
  "runcmd:",
  "  - cd "foo"",
  "  - export RUNNER_ALLOW_RUNASROOT=1",
  "  - >-",
  "    ./config.sh --url https://github.com/owner/repo --token",
  "    githubRegistrationToken --labels label",
  "  - ./run.sh",
  "",
]
`;

exports[`cloud-init with user and ssh key 1`] = `
[
  "#cloud-config",
  "ssh_pwauth: 'no'",
  "users:",
  "  - name: user",
  "    sudo: ALL=(ALL) NOPASSWD:ALL",
  "    shell: /bin/bash",
  "    ssh_authorized_keys:",
  "      - key",
  "runcmd:",
  "  - mkdir actions-runner && cd actions-runner",
  "  - >-",
  "    case $(uname -m) in aarch64) ARCH="arm64" ;; amd64|x86_64) ARCH="x64" ;;",
  "    esac && export RUNNER_ARCH=\${ARCH}",
  "  - >-",
  "    curl -O -L",
  "    https://github.com/actions/runner/releases/download/v2.299.1/actions-runner-linux-\${RUNNER_ARCH}-2.299.1.tar.gz",
  "  - tar xzf ./actions-runner-linux-\${RUNNER_ARCH}-2.299.1.tar.gz",
  "  - export RUNNER_ALLOW_RUNASROOT=1",
  "  - >-",
  "    ./config.sh --url https://github.com/owner/repo --token",
  "    githubRegistrationToken --labels label",
  "  - ./run.sh",
  "",
]
`;

exports[`cloud-init without disable update 1`] = `
[
  "#!/bin/bash",
  "mkdir actions-runner && cd actions-runner",
  "case $(uname -m) in aarch64) ARCH="arm64" ;; amd64|x86_64) ARCH="x64" ;; esac && export RUNNER_ARCH=\${ARCH}",
  "curl -O -L https://github.com/actions/runner/releases/download/v2.299.1/actions-runner-linux-\${RUNNER_ARCH}-2.299.1.tar.gz",
  "tar xzf ./actions-runner-linux-\${RUNNER_ARCH}-2.299.1.tar.gz",
  "export RUNNER_ALLOW_RUNASROOT=1",
  "./config.sh --url https://github.com/owner/repo --token githubRegistrationToken --labels label",
  "./run.sh",
]
`;
